### PR TITLE
Include range of xml docs in declarations and bindings ranges

### PR DESCRIPTION
**Pull request by DedSec256**
We want to be able to analyze comment blocks for declarations/bindings.
This PR suggests adding XmlDoc range to the type declarations/bindings range.

### CODE REVIEW

File: src/fsharp/SyntaxTreeOps.fs [hunk 1]
```diff
@@ -566,9 +569,10 @@ let mkSynBindingRhs staticOptimizations rhsExpr mRhs retInfo =
         | None -> rhsExpr, None
     rhsExpr, retTyOpt
 
-let mkSynBinding (xmlDoc, headPat) (vis, isInline, isMutable, mBind, spBind, retInfo, mEquals, origRhsExpr, mRhs, staticOptimizations, attrs, memberFlagsOpt) =
+let mkSynBinding (xmlDoc: PreXmlDoc, headPat) (vis, isInline, isMutable, mBind, spBind, retInfo, mEquals, origRhsExpr, mRhs, staticOptimizations, attrs, memberFlagsOpt) =
     let info = SynInfo.InferSynValData (memberFlagsOpt, Some headPat, retInfo, origRhsExpr)
     let rhsExpr, retTyOpt = mkSynBindingRhs staticOptimizations origRhsExpr mRhs retInfo
+    let mBind = unionRangeWithXmlDoc xmlDoc mBind
```
Comment (dsyme) [inline src/fsharp/SyntaxTreeOps.fs:575]: Could you explain more - why should the range of the XML docs be included in the range for the binding?

Comment (DedSec256) [reply src/fsharp/SyntaxTreeOps.fs:575]: We want to be able to analyze comment blocks for declarations/bindings in a unified manner. It is most convenient not to calculate the beginning of the declaration/binding depending on the presence/absence of XmlDoc, but to know that XmlDoc is part of the binding, that is the range of XmlDoc is included in the binding range.

Comment (dsyme) [reply src/fsharp/SyntaxTreeOps.fs:575]: I think it's likely that other parts of the compilation logic and tests (e.g. error messages, maybe debug points, editor tooling) depend on the range not including the comment

Comment (DedSec256) [reply src/fsharp/SyntaxTreeOps.fs:575]: At the moment, attributes [are included](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWQDaAPAHKsALnwCuNGgD4AugB0ATjOQ1YgsAA8wAXjABmSCAC%2BQA) in the top binding range. Could this be an argument for including XmlDoc in range as well? (:

And do you think there is a neat way to add XmlDoc to the range for local bindings?

**Comment by dsyme** [comment]
> At the moment, attributes are included in the top binding range. Could this be an argument for including XmlDoc in range as well? (:

Yes, it's a reasonable argument. It may be most relevant for collapsing definitions.

What tests are failing once this change is made?

**Comment by DedSec256** [comment]
it looks like the test StructureTests.``XML doc comments`` has failed. This test takes all elements with multiline range and checks the range/collapse range for them. I displayed graphically expected/actual (green = range, red = collapse range). We may notice that the collapse range does not include the XmlDocs.

Expected
image

Actual
image


there are two more failed tests, but they are insignificant

**Comment by dsyme** [comment]
> I displayed graphically expected/actual (green = range, red = collapse range).

So does this test need fixing, or does your code need fixing?  Thanks!

**Comment by DedSec256** [comment]
> So does this test need fixing, or does your code need fixing?

By analogy with C# and based on how collapsing with attributes currently works, declaration/binding collapsing should be done separately from XML docs collapsing.  So this test needs fixing.

File: tests/service/Symbols.fs [hunk 2]
```diff
@@ -1420,7 +1420,7 @@ type Foo =
                 """
 module internal FSharp.Compiler.ParseHelpers
 
-/// The error raised by the parse_error_rich function, which is called by the parser engine
+// The error raised by the parse_error_rich function, which is called by the parser engine
```
Comment (auduchinok) [inline tests/service/Symbols.fs:1423]: @DedSec256 Why this change was needed?

Comment (DedSec256) [reply tests/service/Symbols.fs:1423]: This test checks that the attributes should be included in the declaration range, so XmlDoc in it seems redundant. We have separate tests with XmlDoc's and attributes in XmlDocTests.fs

File: src/fsharp/XmlDoc.fs [hunk 3]
```diff
@@ -180,11 +172,22 @@ type XmlDocCollector() =
         currentGrabPointCommentsCount <- currentGrabPointCommentsCount + 1
 
     member x.LinesBefore grabPointPos =
-        let lines = savedLinesAsArray.Force()
         match savedGrabPoints.TryGetValue grabPointPos with
-        | true, struct(startIndex, endIndex, _) -> lines.[startIndex .. endIndex]
+        | true, struct(startIndex, endIndex, _) ->
+            let linesBefore = Array.create (endIndex - startIndex + 1) ("", range0)
```
Comment (dsyme) [inline src/fsharp/XmlDoc.fs:177]: Why not use a slice? Isn't this equivalent to `savedLines.[startIndex .. endIndex]`?

**Review by dsyme — CHANGES_REQUESTED** [review]
Just one minor change requested

File: src/fsharp/XmlDoc.fs (hunk 3 above)
Comment (DedSec256) [reply src/fsharp/XmlDoc.fs:177]: `savedLines` is ResizeArray, and it doesn't support slicing :(

image
image

Comment (DedSec256) [reply src/fsharp/XmlDoc.fs:177]: Also, `List<T>` contains `List<T>.GetRange(Int32, Int32)` method, but it returns new `List<T>`, which is not very convenient for further processing

Comment (dsyme) [reply src/fsharp/XmlDoc.fs:177]: I see, thanks

**Comment by dsyme** [comment]
@auduchinok Could I ask for another review on this please?  Do you approve?

File: tests/service/XmlDocTests.fs [hunk 4]
```diff
@@ -10,10 +10,119 @@ module Tests.Service.XmlDocTests.XmlDoc
 open FSharp.Compiler.CodeAnalysis
 open FSharp.Compiler.Service.Tests.Common
 open FSharp.Compiler.Symbols
+open FSharp.Compiler.Syntax
 open FSharp.Test.Compiler
 open FsUnit
 open NUnit.Framework
-open FSharp.Compiler.Diagnostics
+
+let (|Types|TypeSigs|) = function
```
Comment (auduchinok) [inline tests/service/XmlDocTests.fs:18]: @DedSec256 I think it would probably be good to have these active patterns extracted into some common place. There's nothing xmldoc-specific in them, and it could be great to have util code to reuse it in other parser-related tests.

File: tests/service/XmlDocTests.fs [hunk 5]
```diff
@@ -10,10 +10,119 @@ module Tests.Service.XmlDocTests.XmlDoc
...
+        LetBindings(range, bindings)
+
+    | ParsedInput.ImplFile(ParsedImplFileInput(modules = [
+            SynModuleOrNamespace.SynModuleOrNamespace(decls = [
+                SynModuleDecl.DoExpr(expr = SynExpr.LetOrUse(range = range; bindings = bindings))])])) ->
+        LetBindings(range, bindings)
+
+    | ParsedInput.ImplFile(ParsedImplFileInput(modules = [
+            SynModuleOrNamespace.SynModuleOrNamespace(decls = decls)])) ->
+        Lets(decls)
```
Comment (auduchinok) [inline tests/service/XmlDocTests.fs:118]: `Lets` doesn't sound like an suitable name here. Please, consider the following code which doesn't contain any `let` bindings:

```fsharp
module M =
    type T() = class end

    do ()

    open System
```

**Review by auduchinok — COMMENTED** [review]
The parser changes look good to me.

Just a few things about the added test utils.

**Comment by auduchinok** [comment]
@dsyme I think this is good to go in now.

We should probably consider unifying util functions, active patterns and manual pattern matching that are used in various parser/range-related PRs, but this is orthogonal to changes in this PR.

**Review by dsyme — APPROVED** [review]
